### ISSUE TITLE
twisted.web.http.Request allows you to write to a finished request.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4317]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)| @dreid reported|
|-|-|
|Trac ID|trac#4317|
|Type|defect|
|Created|2010-02-24 21:23:14Z|
|Branch|https://github.com/twisted/twisted/tree/request-write-after-finish-4317|

This can cause corruption to subsequent requests, it is also partially responsible for the behavior described in [#3684](https://github.com/twisted/twisted/issues/3684)

Attachments:

* [ticket 4317 patch.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ba/3ba27d9d05dd2bbdd33e81cdc4b8ab6aa14da0e0/bbc1a586c39d25e6fadae1183785371d6ed94304.patch) (380 bytes) - added by hderms on 2011-01-14 20:06:19Z - Simple fix, uses pythonic idiom that was used in the same class
* [ticket 4317 patch v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ba/3ba27d9d05dd2bbdd33e81cdc4b8ab6aa14da0e0/cc9c917f70fb081533401cfbe690dd62c7b0f7f5.patch) (1978 bytes) - added by hderms on 2011-01-25 20:12:39Z - Rewrite of patch and addition of unit test
* [ticket 4317 patch v3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ba/3ba27d9d05dd2bbdd33e81cdc4b8ab6aa14da0e0/8fb5556c9b1e48da2c69c2bc70f7751bbf2e3c35.patch) (1669 bytes) - added by hderms on 2011-02-04 18:51:15Z - adjusted changes requested by reviewer

<details><summary>Searchable metadata</summary>

```
trac-id__4317 4317
type__defect defect
reporter__dreid dreid
priority__normal normal
milestone__ 
branch__branches_request_write_after_finish_4317 branches/request-write-after-finish-4317
branch_author__hderms hderms
status__closed closed
resolution__fixed fixed
component__web web
keywords__easy easy
time__1267046594000000 1267046594000000
changetime__1297302552000000 1297302552000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

We should change `Request.write` to discard such bytes.  We should also introduce a warning for this case (that could be done separately though).  And ultimately, this should result in an exception.

**Comment by twisted-trac** [comment]
|<img alt="hderms's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|hderms removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to hderms|
|-|-|

Thanks.  We need a unit test which verifies that this behaves as desired, too.  Can you add that to the patch?

**Comment by twisted-trac** [comment]
|<img alt="hderms's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|hderms commented|
|-|-|

I will do so. I was confused and thought unit-tests were only required when adding new features. Will read the twisted documentation on unit tests and then implement.

**Comment by twisted-trac** [comment]
|<img alt="hderms's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|hderms removed owner|
|-|-|

Added a new patch which raises a RuntimeError when Request.write is called after Request.finish. Added a unit test which tests this functionality.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to hderms|
|-|-|

 * `+        if self.finished == 1:` You should simply call `if self.finished:' here. This ought to be a boolean, if though it's not...
 * `twisted.web.test.test_xml.MicroDOMTest.testLaterCloserTable` fails after your fix is applied. I suspect the test is broken, but it should be fixed.
 * 
```
-        self.assertRaises(RuntimeError, req.finish)
+        self.assertRaises(RuntimeError, req.finish())
```
  This change is broken, it makes the test fails.
 * There should be 2 blank lines before your test
 * The test docstring looks a bit weird. Is there an extra "L{Request.finish}" at the beginning?

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="hderms's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|hderms commented|
|-|-|

Replying to [therve](#note_6):
>  * `+        if self.finished == 1:` You should simply call `if self.finished:' here. This ought to be a boolean, if though it's not...
>  * `twisted.web.test.test_xml.MicroDOMTest.testLaterCloserTable` fails after your fix is applied. I suspect the test is broken, but it should be fixed.
>  * 
> ```
> -        self.assertRaises(RuntimeError, req.finish)
> +        self.assertRaises(RuntimeError, req.finish())
> ```
>   This change is broken, it makes the test fails.
>  * There should be 2 blank lines before your test
>  * The test docstring looks a bit weird. Is there an extra "L{Request.finish}" at the beginning?
> 
> Thanks!
I examined that test that fails, and I can't seem to find any connection between it and /web/http, it doesn't import anything http related, and the test in particular that fails is just parsing strings. I'm not entirely sure but I think the test is broken as well.


I didn't implement the if self.finished == 1: and I found that to be rather unpythonic as well, but was reluctant to change it because this is my first patch. 

The change to the the test_finishAfterConnectionLost(self): was unintentional. I think it resulted from me examining how these unit tests worked, I must have forgotten to eliminate that portion. I will remove it and resubmit.

I will fix the formatting errors.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

Hum sorry, the test I mentioned  (`twisted.web.test.test_xml.MicroDOMTest.testLaterCloserTable`) was probably just a output parsing error on my part. It must have been the other one.

**Comment by twisted-trac** [comment]
|<img alt="hderms's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|hderms removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [30650]) Branching to 'request-write-after-finish-4317'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [30651]) Apply patch from the ticket, with a couple trivial changes

refs [#4317](https://github.com/twisted/twisted/issues/4317)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [30652]) news fragment

refs [#4317](https://github.com/twisted/twisted/issues/4317)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Did you manually edit the patch to remove the last hunk?  However the latest version of the patch was created, neither patch nor emacs is happy with it (Emacs says "Hunk seriously messed up").  It's not a problem because this patch is short enough that I can manually apply it.  Whatever happened though, try to avoid it in the future. :)

I checked the changes into a branch, added a news fragment, and made the `if self.finished:` change therve suggested.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/request-write-after-finish-4317) look good, so I'll go ahead and merge this.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [30653]) Merge request-write-after-finish-4317

Author: hderms
Reviewer: therve, exarkun
Fixes: [#4317](https://github.com/twisted/twisted/issues/4317)

In `Request.write`, check to see if the response has already been finished and raise
an exception if it has, rather than tossing garbage bytes onto the connection.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
